commit 1d3a8d3160418abca25c876b42db6a368295ae2a
Author: Dimitar Gamishev <gamishev@gmail.com>
Date:   Wed Mar 27 14:41:27 2019 +0200

    Added rtp support

diff --git a/arch/arm/boot/dts/sun5i-a13-olinuxino.dts b/arch/arm/boot/dts/sun5i-a13-olinuxino.dts
index 9409c232d48a..68b9c34861cd 100644
--- a/arch/arm/boot/dts/sun5i-a13-olinuxino.dts
+++ b/arch/arm/boot/dts/sun5i-a13-olinuxino.dts
@@ -48,6 +48,7 @@
 
 #include <dt-bindings/gpio/gpio.h>
 #include <dt-bindings/input/input.h>
+#include <dt-bindings/pwm/pwm.h>
 
 / {
 	model = "Olimex A13-Olinuxino";
@@ -72,43 +73,31 @@
 		};
 	};
 
-	bridge {
-		compatible = "dumb-vga-dac";
-		#address-cells = <1>;
-		#size-cells = <0>;
+	 backlight: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&pwm 0 50000 PWM_POLARITY_INVERTED>;
+                brightness-levels = <0 10 20 30 40 50 60 70 80 90 100>;
+                default-brightness-level = <8>;
 
-		ports {
-			#address-cells = <1>;
-			#size-cells = <0>;
+        };
 
-			port@0 {
-				reg = <0>;
-
-				vga_bridge_in: endpoint {
-					remote-endpoint = <&tcon0_out_vga>;
-				};
-			};
-
-			port@1 {
-				reg = <1>;
-
-				vga_bridge_out: endpoint {
-					remote-endpoint = <&vga_con_in>;
-				};
+	 panel: panel {
+                compatible = "olimex,lcd-olinuxino-4.3";
+                backlight = <&backlight>;
+		enable-gpios = <&axp_gpio 0 GPIO_ACTIVE_HIGH>; /* AXP GPIO0 */
+		pinctrl-names = "default";
+                #address-cells = <1>;
+                #size-cells = <0>;
+		port {
+			panel_input: endpoint {
+				remote-endpoint = <&tcon0_out_lcd>;
 			};
 		};
+
+                };
 	};
 
-	vga {
-		compatible = "vga-connector";
 
-		port {
-			vga_con_in: endpoint {
-				remote-endpoint = <&vga_bridge_out>;
-			};
-		};
-	};
-};
 
 &be0 {
 	status = "okay";
@@ -132,6 +121,11 @@
 
 		interrupt-controller;
 		#interrupt-cells = <1>;
+		   axp_gpio:gpio{ 
+                        compatible = "x-powers,axp209-gpio";
+                        gpio-controller;
+                        #gpio-cells = <2>; 
+                    };
 	};
 };
 
@@ -218,6 +212,8 @@
 	};
 };
 
+
+
 &reg_usb0_vbus {
 	status = "okay";
 	gpio = <&pio 6 12 GPIO_ACTIVE_HIGH>; /* PG12 */
@@ -233,14 +229,19 @@
 	pinctrl-0 = <&lcd_rgb666_pins>;
 	status = "okay";
 };
-
 &tcon0_out {
-	tcon0_out_vga: endpoint@0 {
+	tcon0_out_lcd: endpoint@0 {
 		reg = <0>;
-		remote-endpoint = <&vga_bridge_in>;
+		remote-endpoint = <&panel_input>;
 	};
 };
 
+&pwm {
+        pinctrl-names = "default";
+        pinctrl-0 = <&pwm0_pin>;
+        status = "okay";
+};
+
 &uart1 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&uart1_pg_pins>;
@@ -261,3 +262,12 @@
 	usb1_vbus-supply = <&reg_usb1_vbus>;
 	status = "okay";
 };
+
+&rtp {
+	allwinner,ts-attached;
+	#thermal-sensor-cells = <0>;
+	/* sensitive/noisy touch panel */
+	allwinner,tp-sensitive-adjust = <0>;
+	allwinner,filter-type = <3>;
+};
+
